Firmware/Source/Controller/Commutation.c

```diff
@@ -0,0 +1,177 @@
+void COMM_ApplyCommutation();
+void COMM_CleanBistable();
+
+void COMM_CommutateGroup1();
+void COMM_CommutateGroup2();
+void COMM_DisconnectAllRelay();
+void COMM_ConnectRelayGroup(uint8_t Group);
+void COMM_TransferDataRawToShiftRegistersState();
+void COMM_SetRelayFromRAW();
+void COMM_PrepareFromRAW();
+void COMM_CleanRegisters();
+
+// Functions
+//
+void COMM_SetRelayFromRAW()
+{
+	COMM_PrepareFromRAW();
+	COMM_ApplyCommutation();
+	
+	CONTROL_DelayMs(500);
+	
+	COMM_CleanBistable();
+	COMM_ApplyCommutation();
+}
+// ----------------------------------------
+
+void COMM_ConnectRelayGroup(uint8_t Group)
+{
+	switch (Group)
+	{
+		case 1:
+			{
+				COMM_CommutateGroup1();
+			}
+			break;
+		case 2:
+			{
+				COMM_CommutateGroup2();
+			}
+			break;
+			
+		default:
+			{
+				COMM_DisconnectAllRelay();
+			}
+			break;
+	}
+}
+// ----------------------------------------
+
+void COMM_CommutateGroup1()
+{
+	COMM_SwitchSimpleDevice(COMM_SwitchPOTPlusToPOT1, true);
+	COMM_SwitchBistableDevice(COMM_SwitchBUSLVPlusToPOW1, true);
+	
+	COMM_ApplyCommutation();
+	
+	CONTROL_DelayMs(500);
+	
+	COMM_CleanBistable();
+	COMM_ApplyCommutation();
+}
+// ----------------------------------------
+
+void COMM_CommutateGroup2()
+{
+	COMM_SwitchSimpleDevice(COMM_SwitchPOTPlusToPOT1, true);
+	COMM_SwitchBistableDevice(COMM_SwitchBUSLVPlusToPOW1, true);
+	
+	COMM_ApplyCommutation();
+	
+	CONTROL_DelayMs(500);
+	
+	COMM_CleanBistable();
+	COMM_ApplyCommutation();
+}
+// ----------------------------------------
+
+void COMM_DisconnectAllRelay()
+{
+	COMM_CleanRegisters();
+	ShiftRegistersState[REGISTER_E] = 0xFF;
+	ShiftRegistersState[REGISTER_G] = 0xFF;
+	BistableBits[REGISTER_E] = 0xFF;
+	BistableBits[REGISTER_G] = 0xFF;
+	COMM_ApplyCommutation();
+	
+	CONTROL_DelayMs(500);
+	
+	COMM_CleanBistable();
+	COMM_ApplyCommutation();
+}
+// ----------------------------------------
+
+void COMM_SwitchSimpleDevice(RegisterPin Device, bool State)
+{
+	if(State)
+		ShiftRegistersState[Device.Register] |= Device.Bit;
+	else
+		ShiftRegistersState[Device.Register] &= ~Device.Bit;
+}
+// ----------------------------------------
+
+void COMM_SwitchBistableDevice(BistableSwitch Device, bool State)
+{
+	if(State)
+	{
+		ShiftRegistersState[Device.TurnOnPin.Register] |= Device.TurnOnPin.Bit;
+		BistableBits[Device.TurnOnPin.Register] |= Device.TurnOnPin.Bit;
+	}
+	else
+	{
+		ShiftRegistersState[Device.TurnOffPin.Register] |= Device.TurnOffPin.Bit;
+		BistableBits[Device.TurnOffPin.Register] |= Device.TurnOffPin.Bit;
+	}
+}
+// ----------------------------------------
+
+void COMM_ApplyCommutation()
+{
+	for(uint8_t i = 0; i < REGISTERS_NUM; i++)
+	{
+		SPI_WriteByte(SPI1, ShiftRegistersState[i]);
+	}
+}
+// ----------------------------------------
+
+void COMM_CleanBistable()
+{
+	for(uint8_t i = 0; i < REGISTERS_NUM; i++)
+	{
+		ShiftRegistersState[i] &= BistableBits[i];
+		BistableBits[i] = 0;
+	}
+}
+// ----------------------------------------

```

Comment:
```suggestion
void COMM_CleanBistable()
{
	for(uint8_t i = 0; i < REGISTERS_NUM; i++)
	{
		ShiftRegistersState[i] &= BistableBits[i];
		BistableBits[i] = 0;
	}
}
// ----------------------------------------
void COMM_DisconnectBistable()
{
	for(uint8_t i = 0; i < REGISTERS_NUM; i++)
	{
		ShiftRegistersState[i] &= ~BistableBits[i];
		BistableBits[i] = 0;
	}

	COMM_ApplyCommutation();
}
// ----------------------------------------
```
Зачем эта функция модифицирована? Зачем разделять очистку бистабильных бит и применение этой операции, если они всегда вызываются вместе? Чтобы потом так писать?
```
	COMM_CleanBistable();
	COMM_ApplyCommutation();
```
Дублирование кода и поле для ошибок.
И да, не хватает инверсии `~BistableBits[i]`